src/diagram/models/variable.ts

```diff
@@ -43,117 +43,265 @@ export const Variable = types.model("Variable", {
-      let value;
-      switch (self.operation) {
-        case "÷":
-          // @ts-expect-error THIS
-          value = this.inputA.computedValue / this.inputB.computedValue;
-          break;
-        case "×":
-          // @ts-expect-error THIS
-          value = this.inputA.computedValue * this.inputB.computedValue;
-          break;
-        case "+":
-          // @ts-expect-error THIS
-          value = this.inputA.computedValue + this.inputB.computedValue;
-          break;
-        case "-":
-          // @ts-expect-error THIS
-          value = this.inputA.computedValue - this.inputB.computedValue;
-          break;
-        default:
-          break;
+    // @ts-expect-error THIS
+    const inputBMathValue = this.inputB?.mathValue;
+
+    // The input value could be a number so it could be 0 which would be
+    // valid
+    if (self.inputB && inputBMathValue === undefined) {
+      // @ts-expect-error THIS
+      if (this.inputB.computedValue !== undefined) {
+        // If there is an input value and the math value is not valid that
+        // should mean there is an error in the units
+        return {message: "cannot compute value from inputs"};
+      } else {
+        // If there is no value for this input we cannot compute the output
+        // value. There might be a unit.
+        //
+        // We don't show an explicit error here: The current UI shows NaN for
+        // the value, so this seems like enough of a error message.
+        return {};
+      }
+    }
+
+    try {
+      const result = evaluate(expression, {a: inputAMathValue, b: inputBMathValue});
+      const resultType = typeof result;
+      if (isUnit(result)) {
+        // We need to use simplify here so we are consistent with the unit
+        // calculation. The simplify function will convert the prefix of
+        // units based on the size of the value.
+        const simpl = result.simplify();
+        return {value: simpl.toNumber()};
+      } else if (resultType === "number") {
+        return {value: result};
+      } else {
+        // In theory math.js can return other kinds of results arrays, big
+        // numbers, ...  With the current code that shouldn't be possible
+        // but when we allow expressions it will be more likely to happen
+        return {error: `unknown result type: ${resultType}`};
       }
-      if (self.operation) {
-        return {value};
+    } catch (e: any) {
+      // TODO: we should find a way to handle this without throwing an
+      // exception, but I think that will mean changes to MathJS
+      //
+      // We should update MathJS to provide more information here. For other
+      // errors MathJS provides a data property on the error that includes
+      // the character location and more info about the error.
+      if (e.message?.startsWith("Units do not match")) {
+        return {error: "incompatible units"};
+      } else if (e.message?.startsWith("Unexpected type of argument")) {
```

Comment:
Relying on string contents of error messages seems a bit fragile, but it seems they don't have a numeric error code so it's likely the best we can do. Hopefully, they aren't in the habit of tweaking error message wording on a regular basis. We could write unit tests that verify the error messages if we wanted notification if/when such error messages change but I don't think it need hold up the PR.

Reply:
Yes I agree. I predict we are going to have to work on the error reporting so we can help users figure out what is wrong better. So then we'll work on MathJS. There is an issue in their github repo for this already so I think the change would be welcome.

I had one test for the "Units do not match", but I forgot to add one for the "Unexpected type of argument". It is added now.